### PR TITLE
Add iam managed policies

### LINKED ISSUE #3751: Add support for AWS managed policies
# This is a Feature Proposal

## Description

Currently, I write lambda functions to create AWS accounts, aws roles, policies, ....  To be able to do this, the lambda functions needs almost all IAM actions. I would like to attach AWS managed policies (e.g. IAMFullAccess) instead of define the content of this policy again. Currently I have to add the following lines in the serverless.yml file:
```
iamRoleStatements:
    - Effect: "Allow"
      Action:
        - "iam:*"
      Resource: "*"
```
I'd be grateful if we can attach AWS managed policies instead, for example:
```
iamRoleStatements:
    policies:
        - arn:aws:iam::aws:policy/IAMFullAccess
        - arn:aws:iam::aws:policy/AWSLambdaExecute
```

## Additional Data

* ***Serverless Framework Version you're using***: 1.14.0
* ***Operating System***: WIN7

### PR DESCRIPTION
**Pull request by chrisbarbour**
## What did you implement:

Closes #3751 

## How did you implement it:

I have added an optional section to the aws provider section within the serverless.yml named iamManagedPolicies to go along side iamRoleStatements. It allows users to specify any managed policies at that level without having to explicitly form the whole iam role in the resources section.

## How can we verify it:

Add an iamManagedPolicies section under provider with a list of managed policies like the following serverless.yml (can contain a list of strings or objects to allow for refs and fns in cloudformation):

```yml
service: sandbox-for-managed-policies
provider:
  name: aws
  runtime: java8
  iamManagedPolicies:
    - 'arn:aws:iam:*****:policy/some-managed-policy'
  iamRoleStatements:
    - Effect: Allow
      Action:
        - sns:Publish
      Resource: "*"

package:
  artifact: artifact.jar

functions:
  myFunction:
    handler: someClass
```

This will produce an update stack template with the following IAM role including the added ManagedPolicyArns section:

```json
"IamRoleLambdaExecution": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Effect": "Allow",
              "Principal": {
                "Service": [
                  "lambda.amazonaws.com"
                ]
              },
              "Action": [
                "sts:AssumeRole"
              ]
            }
          ]
        },
        "Policies": [
          {
            "PolicyName": {
              "Fn::Join": [
                "-",
                [
                  "dev",
                  "sandbox-for-managed-policies",
                  "lambda"
                ]
              ]
            },
            "PolicyDocument": {
              "Version": "2012-10-17",
              "Statement": [
                {
                  "Effect": "Allow",
                  "Action": [
                    "logs:CreateLogStream"
                  ],
                  "Resource": [
                    {
                      "Fn::Sub": "arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/sandbox-for-managed-policies-dev-myFunction:*"
                    }
                  ]
                },
                {
                  "Effect": "Allow",
                  "Action": [
                    "logs:PutLogEvents"
                  ],
                  "Resource": [
                    {
                      "Fn::Sub": "arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/sandbox-for-managed-policies-dev-myFunction:*:*"
                    }
                  ]
                },
                {
                  "Effect": "Allow",
                  "Action": [
                    "sns:Publish"
                  ],
                  "Resource": "*"
                }
              ]
            }
          }
        ],
        "Path": "/",
        "ManagedPolicyArns": [
            "arn:aws:iam:*****:policy/some-managed-policy"
        ],
        "RoleName": {
          "Fn::Join": [
            "-",
            [
              "sandbox-for-managed-policies",
              "dev",
              "us-east-1",
              "lambdaRole"
            ]
          ]
        }
      }
    }
```

## Todos:

- [X] Write tests
- [x] Write documentation
- [X] Fix linting errors
- [X] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [X] Enable "Allow edits from maintainers" for this PR
- [X] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: lib/plugins/aws/package/lib/mergeIamTemplates.js [hunk 1]
```diff
@@ -173,4 +188,14 @@ module.exports = {
       throw new this.serverless.classes.Error(errorMessage);
     }
   },
+
+  validateManagedPolicies(iamManagedPolicies) {
+    // Verify that iamManagedPolicies (if present) is an array
+    if (!iamManagedPolicies) {
+      return;
+    }
+    if (!(iamManagedPolicies instanceof Array)) {
```
Comment (horike37) [inline lib/plugins/aws/package/lib/mergeIamTemplates.js]: Use lodash like this `if (!_.isArray(iamManagedPolicies)) {`

File: lib/plugins/aws/package/lib/mergeIamTemplates.js [hunk 2]
```diff
@@ -120,6 +121,20 @@ module.exports = {
           .Statement.concat(this.serverless.service.provider.iamRoleStatements);
     }
 
+    if (this.serverless.service.provider.iamManagedPolicies) {
+      // add iam managed policies
+      const iamManagedPolicies = this.serverless.service.provider.iamManagedPolicies;
+      const resource = this.serverless.service.provider.compiledCloudFormationTemplate
+        .Resources[this.provider.naming.getRoleLogicalId()];
+      if (iamManagedPolicies.length > 0) {
+        if (!resource.ManagedPolicyArns) {
```
Comment (horike37) [inline lib/plugins/aws/package/lib/mergeIamTemplates.js]: Use lodash `if (!_.has(resource, 'ManagedPolicyArns') || _.isEmpty(resource.ManagedPolicyArns)) {`

**Comment by chrisbarbour** [comment]
Thanks @horike37, I have updated with your suggestions

**Comment by horike37** [comment]
Thank you for the update @chrisbarbour :+1: 
I just tested it locally but unfortunately ran into the following error. 

```
  Error --------------------------------------------------
 
  The CloudFormation template is invalid: Invalid template resource property 'ManagedPolicyArns'
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless
 
  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           8.8.1
     Serverless Version:     1.26.0

```

I wonder why this error occurs since `ManagedPolicyArns` is a suitable property as far as reading the documentation. Does this working good on your end?
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html#cfn-iam-role-managepolicyarns

**Comment by chrisbarbour** [comment]
Looks like the ManagedPolicyArns where actually at the wrong position, they need to go to the Properties section, I am updating now and have updated the example above to be correct.
I have verified the update through cloudformation and seems to be it fixed.

**Comment by horike37** [comment]
@chrisbarbour 
Thank you for fixing! It works great at this time!
LGTM :+1:

**Comment by ddougans** [comment]
Hello @chrisbarbour @horike37 this works well until VPC config is added. Once added, the managed policy specified in iamManagedPolicies doesn't appear in the list of attached policies for the IAM role. I wondered if adding AWSLambdaVPCAccessExecutionRole caused the iamManagedPolicies merges to be overwritten. Apologies for not providing at least a test but I am not a js programmer.

**Comment by chrisbarbour** [comment]
Thanks @ddougans. I have opened issue #4876. I will take a look at this asap.

**Pull request merged**